libcaf_net/test/transport_worker.cpp

```diff
@@ -0,0 +1,222 @@
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#define CAF_SUITE transport_worker
+
+#include "caf/net/transport_worker.hpp"
+
+#include "caf/test/dsl.hpp"
+
+#include "host_fixture.hpp"
+
+#include "caf/byte.hpp"
+#include "caf/detail/scope_guard.hpp"
+#include "caf/make_actor.hpp"
+#include "caf/net/actor_proxy_impl.hpp"
+#include "caf/net/endpoint_manager.hpp"
+#include "caf/net/multiplexer.hpp"
+#include "caf/serializer_impl.hpp"
+#include "caf/span.hpp"
+
+using namespace caf;
+using namespace caf::net;
+
+namespace {
+
+constexpr string_view hello_test = "hello test!";
+
+struct application_result {
+  bool initialized;
+  std::vector<byte> data_buffer;
+  std::string resolve_path;
+  actor resolve_listener;
+  atom_value timeout_value;
+  uint64_t timeout_id;
+  sec err;
+};
+
+struct transport_result {
+  std::vector<byte> packet_buffer;
+  ip_endpoint ep;
+};
+
+class dummy_application {
+public:
+  dummy_application(std::shared_ptr<application_result> res)
+    : res_(std::move(res)){
+      // nop
+    };
+
+  ~dummy_application() = default;
+
+  template <class Parent>
+  error init(Parent&) {
+    res_->initialized = true;
+    return none;
+  }
+
+  template <class Transport>
+  void write_message(Transport& transport,
+                     std::unique_ptr<endpoint_manager::message> msg) {
+    transport.write_packet(span<byte>{}, msg->payload);
+  }
+
+  template <class Parent>
+  void handle_data(Parent&, span<const byte> data) {
+    auto& buf = res_->data_buffer;
+    buf.clear();
+    buf.insert(buf.begin(), data.begin(), data.end());
+  }
+
+  template <class Parent>
+  void resolve(Parent&, const std::string& path, actor listener) {
+    res_->resolve_path = path;
+    res_->resolve_listener = std::move(listener);
+  }
+
+  template <class Parent>
+  void timeout(Parent&, atom_value value, uint64_t id) {
+    res_->timeout_value = value;
+    res_->timeout_id = id;
+  }
+
+  void handle_error(sec err) {
+    res_->err = err;
+  }
+
+  static expected<std::vector<byte>> serialize(actor_system& sys,
+                                               const type_erased_tuple& x) {
+    std::vector<byte> result;
+    serializer_impl<std::vector<byte>> sink{sys, result};
+    if (auto err = message::save(sink, x))
+      return err;
+    return result;
+  }
+
+private:
+  std::shared_ptr<application_result> res_;
+};
+
+class dummy_transport {
+public:
+  using transport_type = dummy_transport;
+
+  using application_type = dummy_application;
+
+  dummy_transport(std::shared_ptr<transport_result> res) : res_(res) {
+    // nop
+  }
+
+  template <class Header>
+  void write_packet(Header header, span<const byte> payload, ip_endpoint ep) {
```

Comment:
We know the type of `Header`, no need for a template. 🙂

```suggestion
  void write_packet(span<const byte> header, span<const byte> payload, ip_endpoint ep) {
```